// pragma-views2/app/views/tree-list/data.js
import {RemoteDatasource} from "../../../lib/remote-datasource.js";
import {createLargeFlatListData, createLargeHierarchicalData} from "../virtualization/data.js";
import {HierarchicalItem} from "./../../../lib/hierarchical-item.js";

export function createRootItems() {
    const result = [];
    for (let i = 0; i < 10; i++) {
        const model = {
            "id": `${i}`,
            "title": `Code ${i}`,
            "isFolder": true,
            "expandable": true
        };

        result.push(model);
    }
    return result;
}

function random(max) {
    return Math.floor(Math.random() * Math.floor(max))
}

export function createChildItems(parentItem) {
    const result = [];
    
    let success = true;
    
    for (let i = 0; i < 100; i++) {
        const id = `${parentItem.model.id}.${i}`;
        success = !success;
        
        const model = {
            "id": id,
            "title": `Title ${id}`,
            "code": `Code ${id}`,
            "description": `Description ${id}`,
            "percentage": random(100),
            "success": success,
            "date": new Date()
        }

        result.push(model);
    }
    
    for (let i = 0; i < 10; i++) {
        result[i].success = true;
    }
    
    return result;
}

const schema = {
    datasources: [
        {
            "id": 0,
            "remote": "something",
            "hierarchical": true
        }
    ]
};

export function createDatasource(remoteCallback) {
    const result = new RemoteDatasource({
        definition: schema.datasources[0],
        schema: schema,
        remoteCallback: remoteCallback
    });
    return result;
}
