// pragma-views2/app/views/advanced-list/data-items.js
export function dataItems () {
    const folders = [];
    const assets = [];
    const assetFolders = [];
    const componentsAndTasks = [];
    for (let i = 0; i < 5; i++) {
        const folder = {
            id: i,
            type: "folder",
            description: "Company " + i,
            expandable: true,
            count: 5000
        };
        folders.push(folder);
        for (let j = 0; j < 50; j++) {
            const asset = {
                id: folder.id  +"_"+ j,
                parentId: i,
                type: "asset",
                description: "Asset " + j,
                expandable: true
            };
            assets.push(asset);

            for (let m = 0; m < 2; m++) {
                const componentCount = 15;
                const taskCount = 35;

                const assetFolder = {
                    id: asset.id  +"_"+ m,
                    parentId: asset.id,
                    type: "assetFolder",
                    expandable: true,
                    description: m === 0 ? "Components": "Tasks",
                    subType: m === 0 ? "component": "task",
                    count: m === 0 ? componentCount : taskCount,
                };
                assetFolders.push(assetFolder);

                for (let k = 0; k < componentCount; k++) {
                    const component = {
                        id: assetFolder.id  +"_"+ k,
                        parentId: assetFolder.id,
                        type: "component",
                        description: "Component " + k,
                        failureCount: Math.floor(Math.random() * Math.floor(50)),
                    };
                    componentsAndTasks.push(component);

                }
                for (let l = 0; l < taskCount; l++) {
                    let d = new Date();
                    d.setDate(d.getDate()-l);
                    const task = {
                        id: assetFolder.id  +"_"+ l,
                        parentId: assetFolder.id,
                        type: "task",
                        description: "Task " + l,
                        dateCreated: d.toDateString()
                    };
                    componentsAndTasks.push(task);
                }
            }
        }
    }

    return {
        folders, assets,assetFolders, componentsAndTasks
    };
}

export function flatItems() {
    const items = [];
    for (let i = 0; i < 3000; i++) {
        let item = {
            id: i,
            code: `Code ${i}`,
            description: `Description ${i}`,
            siteCode: `Site Code ${i%10}`,
            region: `Region Code ${i >500? 1: 2}`

        };
        items.push(item);
    }
    return items;
}
