// pragma-views2/webcomponents/lib/mixin-datacollection-state.js
import {ComponentUiStates} from "./../../lib/component-ui-states.js";

export function registerDataCollectionState(obj) {
    obj.dataCollectionStarted = dataCollectionStarted;
    obj.dataCollectionEnded = dataCollectionEnded;
    obj.requestParameterValues = requestParameterValues;
}

export function unregisterDataCollectionState(obj) {
    removeQueryParametersPanel(obj);
    obj.dataCollectionStarted = null;
    obj.dataCollectionEnded = null;
    obj.requestParameterValues = null;
}

async function dataCollectionStarted() {
    if (this.uiState === ComponentUiStates.suspended) return;

    removeQueryParametersPanel(this);
    
    this.loaderUI = document.createElement("pragma-loader-ui");
    this.loaderUI.datasource = this.datasource;
    this.appendChild(this.loaderUI);
}

async function dataCollectionEnded() {
    if (this.loaderUI != null) {
        this.removeChild(this.loaderUI);
        this.loaderUI = null;
    }
}

async function requestParameterValues() {
    if (this.queryParametersPanel != null) return;
    this.queryParametersPanel = document.createElement("pr-query-parameters-panel");
    this.appendChild(this.queryParametersPanel);
    this.queryParametersPanel.targetElement = this;
    this.queryParametersPanel.datasource = this.datasource;
}

function removeQueryParametersPanel(obj) {
    if (obj.queryParametersPanel != null) {
        obj.queryParametersPanel.resetDataSourceFlag();
        
        obj.removeChild(obj.queryParametersPanel);
        obj.queryParametersPanel = null;
    }
}
