// pragma-views2/app/views/dataset-functions/dataset-functions.js
import {ViewBase} from "../../../baremetal/lib/view-base.js";
import {template} from "./schema.js";
import {DatasetFactory} from "./../../../lib/dataset-factory.js";
import {cloneObject} from "../../../baremetal/lib/class-helper.js";

export default class DatasetFunctions extends ViewBase {
    connectedCallback() {
        this.context = this;
        this.schema = cloneObject(template);
        this.init();
        this.schema.variables.routeParameters = {
            meterId: 543
        }
    }
    
    init() {
        const options = {
            schema: this.schema
        };
        
        this.dsFactory = new DatasetFactory(options);
        this.model = this.dsFactory.createDataSet(0);
        this.model.setInitialValues(
            {
                id: 123,
                meterreading: {
                    id: 891
                },
            });
    }
     
    increaseAge() {
        this.model.age++;
    }
    
    update(){
        this.model.update();
    }

    updateReading(){
        this.model.meterreading.update();
    }
}
