// pragma-views2/webcomponents/pragma-advanced-list/navigation.js
import {NavigationBase} from "../lib/navigation-base.js";

export class Navigation extends NavigationBase {

    /***
     * Returns the  horizontal row index position in px for the given index and rowindex.
     * Warning the item must be present in the dom for this method to work, ie. not removed and back in the store.
     * @param {number} index
     * @returns {number}
     */
    getIndexHorizontalPosition(index) {
        const row = this.getRowByIndex(index);
        if (row != null) {
            return row.offsetLeft;
        }
        return 0;
    }

    /***
     * Returns the vertical row index position in px for the given index and rowindex.
     * Warning the item must be present in the dom for this method to work, ie. not removed and back in the store.
     * @param {number} index
     * @returns {number}
     */
    getIndexVerticalPosition(index) {
        return this._instance.dataStore.data.slice(0, index).reduce((a, c) => {
            return a + c.__height;
        }, 0);
    }

    /***
     * @param {number} index
     * @returns {number}
     */
    getIndexWidth(index) {
        const row = this.getRowByIndex(index);
        if (row != null) {
            return row.offsetWidth;
        }
        return 0;
    }

    /***
     * Returns a boolean value indicating whether the index falls within the visible horizontal scroll area.
     * @param {number} index
     * @returns {boolean}
     */
    isIndexInHorizontalView(index) {
        const rowIndexPosition = this.getIndexHorizontalPosition(index);
        return rowIndexPosition >= this.scroller.scrollLeft &&
            rowIndexPosition < this.scroller.scrollLeft + this.scrollerWidth;
    }

    /***
     * Returns a boolean value indicating whether the index falls within the visible vertical scroll area.
     * @param {number} index
     * @returns {boolean}
     */
    isIndexInVerticalView(index) {
        const indexPosition = this.getIndexVerticalPosition(index);
        return indexPosition >= this.scroller.scrollTop &&
            indexPosition <= this.scroller.scrollTop + (this.scrollerHeight - this.getIndexHeight(index));
    }
}
