// pragma-views2/app/views/schema-playground/schema-playground.js
import {ViewBase} from "./../../../baremetal/lib/view-base.js";
import { DatasetFactory } from "./../../../lib/dataset-factory.js";
import { performTriggersFor } from "../../../baremetal/lib/binding/observers.js";

export default class SchemaPlayground extends ViewBase {
    constructor() {
        super();
        this.context = this;
    }

    connectedCallback() {
        this._schemaForm = this._element.querySelector(".schema-form");
        this._textArea = this._element.querySelector(".schema-textarea");
        this._button = this._element.querySelector(".update-schema-button");

        this.registerEvent(this._button, "click", this._click.bind(this));
    }

    disconnectedCallback() {
        this._schemaForm = null;
        this._textArea = null;
        this._button = null;
        this.schema = null;
        this.context = null;
        this.model = null;
        this.dsFactory = null;
    }

    schemaLoaded() {
        const options = {
            schema: this.schema
        };

        this.dsFactory = new DatasetFactory(options);
        this.model = this.dsFactory.createDataSet(0, {}, this.model, "ONE");

        performTriggersFor(this.model);
    }

    _click() {
        this.schema = JSON.parse(this._textArea.value);
    }
}
