// pragma-views2/app/views/complete/data-details.js
export async function getDetailsFor(detailItem, fetchCount) {
    return new Promise(resolve => {
        const max = 10;
        let result = [];
        (function loop(i) {
            if (i < max) {
                setTimeout(() => {
                    result.push(createItem(detailItem, i));
                    if (fetchCount(i) == true) {
                        loop(i+1);
                    }
                    else {
                        resolve(result);
                    }
                }, 0.1)
            }
            else {
                resolve(result);
            }
        })(0);
    })
}

function createItem(parent, index) {
    return {
        id: index,
        detailCode: parent.code,
        code: `Detail Code ${index}`,
        description: `Detail Description ${index}`,
        value: index
    }
}

const dataItems = [
    {
        id: 0,
        code: "detail code 1",
        description: "detail description 1",
        value: 1
    },
    {
        id: 1,
        code: "detail code 2",
        description: "detail description 2",
        value: 2
    },
    {
        id: 2,
        code: "detail code 3",
        description: "detail description 3",
        value: 4
    },
    {
        id: 3,
        code: "detail code 4",
        description: "detail description 4",
        value: 8
    },
    {
        id: 4,
        code: "detail code 5",
        description: "detail description 5",
        value: 16
    }
];
