// pragma-views2/app/views/complete/data-items.js
export const dataItems = [
    {
        id: 0,
        code: "code 1",
        description: "description 1",
        duration: 60,
        make: 1,
        model: 0,
        modifiedOn: "2019-01-28T10:03:11.020Z",
        guaranteeExpiresOn: "2019-01-28" 
    },
    {
        id: 1,
        code: "code 2",
        description: "description 2",
        duration: 10,
        make: 2,
        model: 1,
        modifiedOn: "1999-01-11T10:03:11.020Z",
        guaranteeExpiresOn: "2019-06-12"
    },
    {
        id: 2,
        code: "code 3",
        description: "description 3",
        duration: 365,
        make: 3,
        model: 1,
        modifiedOn: "2019-01-29T10:03:11.020Z",
        guaranteeExpiresOn: "2011-11-16"
    },
    {
        id: 3,
        code: "code 4",
        description: "description 4",
        duration: 0,
        make: 1,
        model: 1,
        modifiedOn: "2016-01-28T22:00:00.000Z",
        guaranteeExpiresOn: "1999-02-29"
    },
    {
        id: 4,
        code: "code 5",
        description: "description 5",
        duration: null,
        make: 2,
        model: 2,
        modifiedOn: "2018-03-11T09:00:00.000Z",
        guaranteeExpiresOn: "2017-12-31"
    }
];
