// pragma-views2/app/views/_error-nested-repeats/_error-nested-repeats.js
import {ViewBase} from "../../../baremetal/lib/view-base.js";
import {template as mainSchema} from "./schema.js";
import {cloneObject} from "../../../baremetal/lib/class-helper.js";
import {DatasetFactory} from "./../../../lib/dataset-factory.js";

export default class ErrorNestedRepeats extends ViewBase {

    connectedCallback() {
        this.title = "Erorr Nested Repeats";
        this.context = this;
        this.mainSchema = cloneObject(mainSchema);
        this.init();
    }

    init() {
        const options = {
            schema: this.mainSchema
        };

        this.dsFactory = new DatasetFactory(options);
        this.model = this.dsFactory.createDataSet(0);
        this.model.setInitialValues({});
    }

    addBatchItem() {
        const item = {
            id: this.model.batchValues._data.length,
            field: "",
            value: ""
        };
        this.model.batchValues._data.push(item);
    }

    removeBatchItem(params, event) {
        const id = event.target.dataset.id;
        const item = this.model.batchValues._data.find(item => item.id == id);
        this.model.batchValues.deleteItem(item);
    }
}
