// pragma-views2/webcomponents/pr-calendar/day.js
import {sameDay} from "../../lib/date-helpers.js";
import {padStart} from "../../lib/string-helpers.js";

export class Day {

    constructor(index, date, isBufferDay) {
        this._index = index;
        this._date = date;
        this._isBufferDay = isBufferDay;
    }

    get aria() {
        return this.date.toDateString();
    }

    // noinspection JSUnusedGlobalSymbols
    get containerClass() {
        let containerClass = "day-container";
        containerClass = this.isBufferDay ? containerClass + " suppressed" : containerClass;
        containerClass = sameDay(this.date, new Date()) ? containerClass + " today" : containerClass;
        return containerClass;
    }

    // noinspection JSUnusedGlobalSymbols
    get dataValue() {
        return `${this.date.getFullYear()}-${padStart(this.date.getMonth() + 1)}-${padStart(this.date.getDate())}`;
    }

    get date() {
        return this._date;
    }

    set date(newDate) {
        this._date = newDate;
    }

    get day() {
        return this.date.getDate();
    }

    get index() {
        return this._index;
    }

    get isBufferDay() {
        return this._isBufferDay;
    }


}
