// pragma-views2/app/views/context-menu/context-menu.js
import {ViewBase} from "../../../baremetal/lib/view-base.js";
import {contextItems} from "./context-items.js";
import {cloneObject} from "./../../../baremetal/lib/class-helper.js";

export default class ContextMenu extends ViewBase {
    connectedCallback() {
        this.contextMenuEventHandler = this._showContextMenu.bind(this);
        this.contextMenuCallbackHandler = this._contextMenuCallback.bind(this);
        
        this.contextMenuContainer = document.getElementById("context-menu-container");
        this.callbackResultsContainer = document.getElementById("callback-results-container");
        
        const options = {
            element: this.contextMenuContainer,
            eventHandler: this.contextMenuEventHandler
        };
        window.eventEmitter.emit("register-context-menu", options);
        
        this.contextMenuContainer.focus();
    }
    
    disconnectedCallback() {
        window.eventEmitter.emit("unregister-context-menu", { element: this.contextMenuContainer });
        this.contextMenuCallbackHandler = null;
        this.contextMenuEventHandler = null;
        this.callbackResultsContainer = null;
        this.contextMenuContainer = null;        
    }

    _contextMenuCallback(menuItem) {
        this.callbackResultsContainer.innerText = `Callback triggered - Menu Item Id [${menuItem.id}] selected`;
    }

    _showContextMenu(event) {
        const options = {
            title: "Custom Title",
            callback: this.contextMenuCallbackHandler,
            menuItems: cloneObject(contextItems),
            event: event
        };
        window.eventEmitter.emit("show-context-menu", options);
    }
}
