// pragma-views2/app/views/_error-pragma-form-set-model/_error-pragma-form-set-model.js
import {ViewBase} from "../../../baremetal/lib/view-base.js";
import {template as mainSchema} from "./schema.js";
import {cloneObject} from "./../../../baremetal/lib/class-helper.js";
import {DatasetFactory} from "./../../../lib/dataset-factory.js";
import {dataItems} from "./data-items.js";
import {performTriggersFor} from "./../../../baremetal/lib/binding/observers.js";

export default class ErrorFormSetModel extends ViewBase {

    connectedCallback() {
        this.title = "Complete";
        this.context = this;
        this.remoteDsCallbackHandler = this.remoteDsCallback.bind(this);
    }

    loadModel() {
        const options = {
            schema: this.mainSchema,
            remoteDsCallback: this.remoteDsCallbackHandler
        };

        this.dsFactory = new DatasetFactory(options);
        this.model = this.dsFactory.createDataSet(0, {}, this.model);
        performTriggersFor(this.model);
    }

    schemaLoaded(event) {
       console.log("Schema loaded");
        this.loadModel();
    }

    set() {
        this.mainSchema = cloneObject(mainSchema);
    }

    remoteDsCallback(def, payload, parent) {
        return dataItems;

    }

}
