// pragma-views2/lib/scene-graph/scene-node.js
export class SceneNode {
    constructor(value, hasSpacing = false) {
        this.value = value;
        this.hasSpacing = hasSpacing;
        this.hasChildren = false;
        this.isActive = false;
        this._iconWidth = 0;
        this.isAction = false;
        this.xOffset = 0;
        this.yOffset = 0;
        return this;
    }

    dispose() {
        this.id = null;
        this.value = null;
        this.hasSpacing = null;
        this.hasChildren = null;
        this.isActive = null;
        this._iconWidth = null;
        this.isAction = null;
        this.xOffset = null;
        this.textWidth = null;
        this.xOffset = null;
        this.yOffset = null;
        this.returnType = null;
    }
    
    get iconName() {
        return this._iconName;
    }
    
    set iconName(newValue) {
        this._iconName = newValue != null ? (newValue.charAt(0) !== "#" ? `#${newValue}` : newValue) : newValue;
    }

    get iconWidth() {
        return this._iconWidth || 0;
    }

    set iconWidth(newValue) {
        this._iconWidth = newValue;
    }

    toString() {
        let result = this.value || "";
        
        if (this.hasSpacing === true) {
            result = ` ${result} `;
        }
        
        return result;
    }
}
