// pragma-views2/webcomponents/pragma-panel-list/pragma-panel-list.js
import {BaseElement} from "./../../baremetal/lib/base-element.js";
import {ComponentUiStates} from "./../../lib/component-ui-states.js";
import {GroupClickBehaviour} from "./../pragma-advanced-list/group-click-behaviour.js";

export class PragmaPanelList extends BaseElement {
    connectedCallback() {
        super.connectedCallback();
        this.initTemplate();
    }

    disconnectedCallback() {
        super.disconnectedCallback();
        this.filter = null;
        this.list = null;
        this._datasource = null;
        this.selectionHandler = null;
    }
    
    get datasource() {
        return this._datasource;
    }

    set datasource(newValue) {
        this._datasource = newValue;
        this.filter.datasource = newValue;
        this.list.datasource = newValue;
    }
    
    /**
     * Initialise template
     */
    initTemplate() {
        const instance = document.importNode(window.templates.get("pragma-panel-list-template"), true);
        this.appendChild(instance);
    }

    /**
     * list property changed event handler
     */
    listChanged() {
        const titleRef = this.getAttribute("title-ref");
        this.list.setAttribute("title-ref", titleRef);

        const typeRef = this.getAttribute("type-ref");
        this.list.setAttribute("type-ref", typeRef);

        const expandRef = this.getAttribute("expand-ref");
        this.list.setAttribute("expand-ref", expandRef);

        const selection = this.getAttribute("selection");
        this.list.setAttribute("selection", selection);

        const template = this.querySelector("template");
        this.list.appendChild(template);
        
        this.list._groupClickBehaviour = GroupClickBehaviour.expand;
        this.list.uiState = ComponentUiStates.active;
        this.list.handleSelection = this.selectionHandler;
    }
}

customElements.define("pragma-panel-list", PragmaPanelList);
