// pragma-views2/app/views/date-time/date-time.js
import {ViewBase} from "./../../../baremetal/lib/view-base.js";

export default class DateTime extends ViewBase {
    connectedCallback() {
        this.person = {
            firstName: "User Name",
            lastName: "Last Name",
            age: 30,
            createdOn: "2019-02-23T12:12:02.123Z",
            guaranteeDate: "2019-02-23",
            startTime: "12:22:44.000"
        }

        // Added it to window for ease of access
        window.simplePerson = this.person;

        this._inputDate = document.getElementById("myInputDate");
        this._inputDate.dateChangedCallback = date => console.log(`Date changed to: ${date}`);
        this._inputTime = document.getElementById("myInputTime");
        this._inputTime.timeChangedCallback = time => console.log(`Time changed to: ${time}`);
    }

    disconnectedCallback(){
        this._inputDate.dateChangedCallback = null;
        this._inputDate = null;
        this._inputTime.timeChangedCallback = null;
        this._inputTime = null;
        window.simplePerson = null;
        delete window.simplePerson;
    }

}
