// pragma-views2/webcomponents/pragma-loader-ui/pragma-loader-ui.js
import {BaseElement} from "./../../baremetal/lib/base-element.js";

export class PragmaLoaderUI extends BaseElement {
    get datasource() {
        return this._datasource;
    }

    set datasource(newValue) {
        // unregister
        if (this._datasource != null) {
            this._datasource.uiStateManager.unregisterComponent(this);
        }

        this._datasource = newValue;

        // register
        if (newValue != null) {
            this._datasource.uiStateManager.registerComponent(this);
        }
    }

    get fetchMore() {
        return this._fetchMore;
    }

    set fetchMore(newValue) {
        this._fetchMore = newValue;
    }

    connectedCallback() {
        super.connectedCallback();
        this.initTemplate();

        this.registerEvent(this.querySelector("button"), "click", this.stop.bind(this));
        this.fetchMore = true;
    }

    disconnectedCallback() {
        this.datasource = null;
        this.textElement = null;
        super.disconnectedCallback();
    }

    initTemplate() {
        const instance = document.importNode(window.templates.get("pragma-loader-ui-template"), true);
        this.appendChild(instance);
        this.textElement = this.querySelector("p");
    }

    stop() {
        this.fetchMore = false;
    }

    fetchingData(count) {
        requestAnimationFrame(() => {
            if (this.textElement != null) {
                this.textElement.innerText = `Records Fetched: ${count}`;
            }
        });

        return this.fetchMore;
    }
}
customElements.define('pragma-loader-ui', PragmaLoaderUI);
